9.trees_and_graphs/task2.go

```diff
@@ -0,0 +1,582 @@
+		n.Swap(y, z)
+	}
+}
+
+// Sort method
+func (n *Node) Sort() {
+	if n.size == 1 {
+		return
+	}
+	if n.size == 2 {
+		n.Sort2(&n.key[0], &n.key[1])
+	}
+	if n.size == 3 {
+		n.Sort3(&n.key[0], &n.key[1], &n.key[2])
+	}
+}
+
+// InsertToNode method insert key `k` to the vertex (not in tree)
+func (n *Node) InsertToNode(k int) {
+	n.key[n.size] = k
+	n.size++
+	n.Sort()
+}
+
+// RemoveFromNode method remove key `k` from the vertex (not from tree)
+func (n *Node) RemoveFromNode(k int) {
+	if n.size >= 1 && n.key[0] == k {
+		n.key[0] = n.key[1]
+		n.key[1] = n.key[2]
+		n.size--
+	} else if n.size == 2 && n.key[1] == k {
+		n.key[1] = n.key[2]
+		n.size--
+	}
+}
+
+// BecomeNode2 method transform to second vertex
+func (n *Node) BecomeNode2(k int, first, second *Node) {
+	n.key[0] = k
+	n.first = first
+	n.second = second
+	n.third = nil
+	n.fourth = nil
+	n.parent = nil
+	n.size = 1
+}
+
+// IsLeaf method whether the node is a leaf; validation is used on insertion and deletion
+func (n *Node) IsLeaf() bool {
+	return n.first == nil && n.second == nil && n.third == nil
+}
+
+// Split func devide a node if it contains 3 keys
+func Split(item *Node) *Node {
+	if item.size < 3 {
+		return item
+	}
+
+	x := &Node{
+		size:   1,
+		key:    [3]int{item.key[0]},
+		first:  item.first,
+		second: item.second,
+		parent: item.parent,
+	}
+	y := &Node{
+		size:   1,
+		key:    [3]int{item.key[2]},
+		first:  item.third,
+		second: item.fourth,
+		parent: item.parent,
+	}
+
+	switch {
+	case x.first != nil:
+		x.first.parent = x
+	case x.second != nil:
+		x.second.parent = x
+	}
+	switch {
+	case y.first != nil:
+		y.first.parent = y
+	case y.second != nil:
+		y.second.parent = y
+	}
+
+	if item.parent != nil {
+		item.parent.InsertToNode(item.key[1])
+
+		switch item {
+		case item.parent.first:
+			item.parent.first = nil
+		case item.parent.second:
+			item.parent.second = nil
+		case item.parent.third:
+			item.parent.third = nil
+		}
+
+		switch {
+		case item.parent.first == nil:
+			item.parent.fourth = item.parent.third
+			item.parent.third = item.parent.second
+			item.parent.second = y
+			item.parent.first = x
+		case item.parent.second == nil:
+			item.parent.fourth = item.parent.third
+			item.parent.third = y
+			item.parent.second = x
+		default:
+			item.parent.fourth = y
+			item.parent.third = x
+		}
+
+		parent := item.parent
+		item = nil
+		return Split(parent)
+	} else {
+		x.parent = item
+		y.parent = item
+		item.BecomeNode2(item.key[1], x, y)
+		return item
+	}
+}
+
+// Insert is func, that insert key `k` to the tree with root `p`
+// Always reterns root of the tree, becouse it was change
+func Insert(p *Node, k int) *Node {
+	if p == nil {
+		return &Node{size: 1, key: [3]int{k}}
+	}
+
+	switch {
+	case p.IsLeaf():
+		p.InsertToNode(k)
+	case k <= p.key[0]:
+		p.first = Insert(p.first, k)
+	case p.size == 1 || (p.size == 2 && k <= p.key[1]):
+		p.second = Insert(p.second, k)
+	default:
+		p.third = Insert(p.third, k)
```

Comment:
вот тут у тебя багуля, ты присваиваешь потомку его родителя, из-за этого происходит зацикливание в дереве, чего допускать нельзя)
```suggestion
	case k <= p.key[0]:
		Insert(p.first, k)
	case p.size == 1 || (p.size == 2 && k <= p.key[1]):
		Insert(p.second, k)
	default:
		Insert(p.third, k)
```